parachain/pallets/inbound-queue/src/lib.rs

```diff
@@ -311,5 +313,20 @@ pub mod pallet {
 			let (xcm_hash, _) = send_xcm::<T::XcmSender>(dest, xcm).map_err(Error::<T>::from)?;
 			Ok(xcm_hash)
 		}
+
+		pub fn decode(log: Log) -> Result<Envelope, Error<T>> {
+			let envelope = Envelope::try_from(log).map_err(|_| Error::<T>::InvalidEnvelope)?;
+			Ok(envelope)
+		}
+
+		pub fn validate(envelope: &Envelope) -> Result<Channel, Error<T>> {
+			// Verify that the message was submitted from the known Gateway contract
+			ensure!(T::GatewayAddress::get() == envelope.gateway, Error::<T>::InvalidGateway);
+
+			// Retrieve the registered channel for this message
+			let channel: Channel =
+				T::ChannelLookup::lookup(envelope.channel_id).ok_or(Error::<T>::InvalidChannel)?;
+			Ok(channel)
+		}
 	}
```

Comment:
I don't think this is an improvement. The previous code was easier to follow, even if the `submit` function was rather long. Its mostly straight-line code without any branching.